### PR TITLE
Refactor event building into EventBuilder

### PR DESCRIPTION
**Pull request by erikjohnston**
This is so that everything is done in one place, making it easier to change the event format based on room version.

The crux of what's happening here is that the `EventBuilder` is now meant to be independent of the format version, allowing the rest of the code to build up events without worrying about different versions. This is done by making it an `attr` class with well defined fields that are of interest when building an event (mainly `type`/`state_key`/`content`/etc), and only adding the rest of the fields when we finally build the event (which handles things like adding prev events, signing and hashing it, using the right format etc)

Hopefully the commits are relatively stand alone.

### CODE REVIEW

File: synapse/handlers/message.py [hunk 1]
```diff
@@ -557,26 +556,12 @@ def create_new_client_event(self, builder, requester=None,
...
-                context.prev_state_events
-            )
-
-        yield self.auth.add_auth_events(builder, context)
-
-        signing_key = self.hs.config.signing_key[0]
-        add_hashes_and_signatures(
-            builder, self.server_name, signing_key
+        event = yield builder.build(
+            prev_event_ids=[p for p, _ in prev_events],
```
Comment (erikjohnston) [inline synapse/handlers/message.py]: The reason for not also passing in the hashes is that we're gonna need to get rid of the concept of "event_id + hashes" in the brave new world

File: synapse/events/builder.py [hunk 2]
```diff
@@ -13,65 +13,124 @@
...
-def get_event_builder(room_version, key_values={}, internal_metadata_dict={}):
-    """Generate an event builder appropriate for the given room version
 
-    Args:
-        room_version (str): Version of the room that we're creating an
-            event builder for
-        key_values (dict): Fields used as the basis of the new event
-        internal_metadata_dict (dict): Used to create the `_EventInternalMetadata`
-            object.
+@attr.s(slots=True, hash=False, frozen=True)
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/events/builder.py]: Bah, somehow thought it disabled allowing being hashed at all.

File: synapse/events/builder.py [hunk 3]
```diff
@@ -13,65 +13,124 @@
...
 class EventBuilderFactory(object):
-    def __init__(self, clock, hostname):
-        self.clock = clock
-        self.hostname = hostname
+    def __init__(self, hs):
+        self.clock = hs.get_clock()
+        self.hostname = hs.hostname
+        self.signing_key = hs.config.signing_key[0]
+
+        self.store = hs.get_datastore()
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/events/builder.py]: `EventBuilder` uses them, but it's probably better to just pass those in directly

**Comment by erikjohnston** [comment]
Now based on #4493 and #4494

File: synapse/events/builder.py [hunk 4]
```diff
@@ -13,79 +13,156 @@
...
-            unsigned=unsigned,
-            internal_metadata_dict=internal_metadata_dict,
+        Args:
+            prev_event_ids (list[str]): The event IDs to use as the prev events
+
+        Returns:
+            Deferred[FrozenEvent]
+        """
+
+        state_ids = yield self._state.get_current_state_ids(
```
Comment (richvdh) [inline synapse/events/builder.py]: Slightly concerned that we used to do `context.get_prev_state_ids`, which might return something different to this?

Comment (erikjohnston) [reply synapse/events/builder.py]: In practice, they're exactly the same, `compute_event_context` just resolves the state across the `prev_events` to get `prev_state_ids`.  From a protocol perspective, `auth_events` should be based on the state at the event.

We could make `compute_event_context` work with an `EventBuilder`, but I was slightly shying away from doing so in the interests of not changing too much. We could probably do so though

**Pull request merged**